internal/command/views/cloud.go

```diff
@@ -4,53 +4,192 @@
+var _ Cloud = (*CloudHuman)(nil)
+
+func (v *CloudHuman) Diagnostics(diags tfdiags.Diagnostics) {
+	v.view.Diagnostics(diags)
+}
+
+func (v *CloudHuman) RetryLog(attemptNum int, resp *http.Response) {
+	msg, elapsed := retryLogMessage(attemptNum, resp, &v.lastRetry)
+	// retryLogMessage returns an empty string for the first attempt or for rate-limited responses (HTTP 429)
+	if msg != "" {
+		if elapsed != nil {
+			v.Output(msg, elapsed) // subsequent retry message
+		} else {
+			v.Output(msg)            // initial retry message
+			v.view.streams.Println() // ensures a newline between messages
+		}
 	}
+}
 
-	var msg string
-	if attemptNum == 1 {
-		msg = initialRetryError
-	} else {
-		msg = fmt.Sprintf(repeatedRetryError, time.Since(hooks.lastRetry).Round(time.Second))
+func (v *CloudHuman) Output(messageCode CloudMessageCode, params ...any) {
+	v.view.streams.Println(v.PrepareMessage(messageCode, params...))
+}
+
+func (v *CloudHuman) PrepareMessage(messageCode CloudMessageCode, params ...any) string {
+	message, ok := CloudMessageRegistry[messageCode]
+	if !ok {
+		// display the message code as fallback if not found in the message registry
+		return string(messageCode)
 	}
 
-	if colorize {
-		return strings.TrimSpace(fmt.Sprintf("[reset][yellow]%s[reset]", msg))
+	if message.HumanValue == "" {
+		// no need to apply colorization if the message is empty
+		return message.HumanValue
 	}
-	return strings.TrimSpace(msg)
+
+	return v.view.colorize.Color(strings.TrimSpace(fmt.Sprintf(message.HumanValue, params...)))
 }
 
-// The newline in this error is to make it look good in the CLI!
-const initialRetryError = `
+// The CloudJSON implementation renders streaming JSON logs, suitable for
+// integrating with other software.
+type CloudJSON struct {
+	view *JSONView
+
+	lastRetry time.Time
+}
+
+var _ Cloud = (*CloudJSON)(nil)
+
+func (v *CloudJSON) Diagnostics(diags tfdiags.Diagnostics) {
+	v.view.Diagnostics(diags)
+}
+
+func (v *CloudJSON) RetryLog(attemptNum int, resp *http.Response) {
+	msg, elapsed := retryLogMessage(attemptNum, resp, &v.lastRetry)
+	// retryLogMessage returns an empty string for the first attempt or for rate-limited responses (HTTP 429)
+	if msg != "" {
+		if elapsed != nil {
+			v.Output(msg, elapsed) // subsequent retry message
+		} else {
+			v.Output(msg) // initial retry message
+		}
+	}
+}
+
+func (v *CloudJSON) Output(messageCode CloudMessageCode, params ...any) {
+	// don't add empty messages to json output
+	preppedMessage := v.PrepareMessage(messageCode, params...)
+	if preppedMessage == "" {
+		return
+	}
+
+	current_timestamp := time.Now().UTC().Format(time.RFC3339)
+	json_data := map[string]string{
+		"@level":       "info",
+		"@message":     preppedMessage,
+		"@module":      "terraform.ui",
+		"@timestamp":   current_timestamp,
+		"type":         "cloud_output",
+		"message_code": string(messageCode),
+	}
+
+	cloud_output, _ := json.Marshal(json_data)
+	v.view.view.streams.Println(string(cloud_output))
```

Comment:
Would this be an appropriate solution otherwise?

```go
	cloud_output, err := json.Marshal(json_data)
	if err != nil {
		v.view.view.streams.Println("Error marshalling JSON:", err)
		return
	}

	v.view.view.streams.Println(string(cloud_output))
```

Reply:
Thanks for the suggestion, I've always applied this change to the implementation in `init.go`. See 6b0ba4d.